### ISSUE TITLE
Using rem as unit in transforms

### ISSUE THREAD

**Issue by k1sul1** [react-grid-layout/react-draggable#480]
Draggable works just as I need, with one exception. The CSS transforms are done with pixels, but I need them to be rems, as my application uses them for scaling stuff down 1:1. Pretty much everything in my app with a pixel unit is converted with this little helper: (x) => `${x / 16}rem`

Is it possible to change those without ditching the batteries and using DraggableCore? I took a quick peek at the source and looks like no, this requires some changes. 

```typescript

export function getTranslation({x, y}: ControlPosition, positionOffset: PositionOffsetControlPosition, unitSuffix: string): string {
  let translation = `translate(${x}${unitSuffix},${y}${unitSuffix})`;
  if (positionOffset) {
    const defaultX = `${(typeof positionOffset.x === 'string') ? positionOffset.x : positionOffset.x + unitSuffix}`;
    const defaultY = `${(typeof positionOffset.y === 'string') ? positionOffset.y : positionOffset.y + unitSuffix}`;
    translation = `translate(${defaultX}, ${defaultY})` + translation;
  }
  return translation;
}
```

Maybe with a `customizer` that takes the translation and params used to build it, and does whatever it wants with it? Not everyone is using 16 as the baseline value so it might be hard to support rems directly, so this might be an ok compromise. 

```typescript

export function getTranslation({x, y}: ControlPosition, positionOffset: PositionOffsetControlPosition, unitSuffix: string, customizer?: (translation: string, xy: ControlPosition, offset: PositionOffsetControlPosition, unitSuffix: string) => string): string {
  let translation = `translate(${x}${unitSuffix},${y}${unitSuffix})`;

  if (customizer) {
    return customizer(translation, { x, y }, positionOffset, unitSuffix)
  }

  if (positionOffset) {
    const defaultX = `${(typeof positionOffset.x === 'string') ? positionOffset.x : positionOffset.x + unitSuffix}`;
    const defaultY = `${(typeof positionOffset.y === 'string') ? positionOffset.y : positionOffset.y + unitSuffix}`;
    translation = `translate(${defaultX}, ${defaultY})` + translation;
  }
  return translation;
}
```

I haven't tested this, just wrote it inline. Is there an easier / better / already supported way?

I tried using the scale property, but either I'm using it wrong or it doesn't work in my case. ATM I have my application scaled to ~13px so everything on the "canvas" fits the screen. That means that items on the canvas (draggable) are 0.8102232667450059% of their full size, and I used that as the scale value. The item is supposed to be at the bottom of the canvas, but instead of that, it's outside, not visible. If I scale my app to 100%, it's where it's supposed to be. 

If I manually convert the pixels in the transform to rem, everything works as expected. So I'd just like to use rem, if that's possible. 

I don't know if my use case is somewhat special or not. It's a "design your own product" kind of app.  Instead of working top-left, my app works bottom-center, meaning [0,0] is dead center on the canvas bottom axis. That way I don't have to deal with repositioning everything when screen size changes, or calculate where the center is. If I want an element 200px on the left, I set x to -200. If I want it 50px from the bottom, I set y to 50. Don't know if that messes with the internal calculations of this component, but FYI.

**Comment by STRML** [comment]
Yeah - your use case makes sense, and I'd love to be able to support it with a stable API and without requiring forking. Doing so would require refactoring such that the major calculations (such as `getTranslation` itself) had a default implementation but could be overridden via props.

I don't really have a good story for this yet, because doing so would require some very deliberate API design; if we just let users override anything they like via props, we're sure to cause breakage, all the time, as we release future minor versions. I'm open to suggestion here.

**Comment by STRML** [comment]
Now, rather insanely enough, you *could* get around this by defining a child component that always translates its `style` attribute, whenever it is passed, to `rem` from `px`, like so:

```jsx
class RemWrapper {
  render() {
    const child = React.Children.only(this.props.children);
    const style = {
      ...this.props.style,
      transform: transformCustomizer(this.props.style.transform),
    };
    return React.cloneElement(child, {style});
  }
}
```

**Comment by k1sul1** [comment]
Thanks for the reply. 

If you think your fix was insane, just look at this. 
~~For now, I can get by with this monster. I'd rather kill it, but I also gotta ship a demo this week.~~

```
document.querySelectorAll('.react-draggable').forEach(el => {
  if (el.style.transform) {
     const containsPx = el.style.transform.indexOf('px') > 0

     if (containsPx) {
       const [x, y] = el.style.transform.replace('translate(', '').replace(')', '').split(', ').map(pxd => pxd.replace('px', '')).map(px => parseInt(px, 10) / 16).map(x => `${x}rem`)
       const newTransform = `translate(${x}, ${y})`

       el.style.transform = newTransform
       console.log(el)
     }
  }
})
```
I'll try your way tomorrow, I've been up way too long already. If it doesn't work, I can always just fork the repo and do the little customizer thing to get it working ~now, and ditch it if the library starts supporting them at some point in time.

**Comment by STRML** [comment]
Fair enough. I'm not against the customizer in principle, but I believe there are more things that users will want to customize than just this - and if we do it, I would like a unified API that I'm prepared to not break between releases.

**Comment by k1sul1** [comment]
Yes, I agree. I just have my tunnel vision goggles on as a dumb user :wink: 

I'm OK with any reasonable solution, if RemWrapper works, that's perfect for me. I'll close the issue if it does, and maybe send a little PR documenting it's use?

**Comment by STRML** [comment]
Sounds good, thanks.

I like your personal site btw - nice work.

**Comment by k1sul1** [comment]
Looks like it works. I haven't done extensive testing yet but at least things render in the right place, and dragging seems to work! It even scales down beautifully when I resize the window.

I had to ditch TS because it wouldn't stop bugging me about this. 

>  TS2605: JSX element type 'RemWrapper' is not a constructor function for JSX elements.
  Types of property 'render' are incompatible.
    Type '({ children, style: styles }: { children: any; style: any; }) => DetailedReactHTMLElement<{ style: any; }, HTMLElement>' is not assignable to type '() => ReactNode'.

I also had to make some changes to RemWrapper to get it to work. I know you probably wrote it as pseudo. I have no idea why this works, but it does. 

```javascript
import React from 'react'
import PropTypes from 'prop-types'

class RemWrapper extends React.Component {
  translateTransformToRem(transform, remBaseline = 16) {
    const convertedValues = transform.replace('translate(', '').replace(')', '')
      .split(',')
      .map(px => px.replace('px', ''))
      .map(px => parseInt(px, 10) / remBaseline)
      .map(x => `${x}rem`)
    const [x, y] = convertedValues

    return `translate(${x}, ${y})`
  }

  render() {
    const { children, remBaseline = 16, style } = this.props
    const child = React.Children.only(children)

    const editedStyle = {
      ...child.props.style,
      ...style,
      transform: this.translateTransformToRem(style.transform, remBaseline),
    }

    return React.cloneElement(child, {
       ...child.props,
       ...this.props,
       style: editedStyle
    })
  }
}

RemWrapper.propTypes = {
  style: PropTypes.shape({
    transform: PropTypes.string.isRequired
  }),
  children: PropTypes.node.isRequired,
  remBaseline: PropTypes.number,
}

export default RemWrapper
```

**Comment by k1sul1** [comment]
If that looks ok to you, I can add a section to the README with an example usage. It assumes that the transform is always `translate(x, y)`, and will break if it encounters anything else. It's fine for me, but it may have some unintended consequences for someone else. 

Also thanks, you're like the 20th person to visit my site! I created that about a year ago just before I became a full-time freelancer, and didn't have anything to do for about 3 weeks. I haven't been able to touch it after that... 

It's an interesting combination. Nginx routes known bot useragents to a "static" Puppeteer-rendered version of the site so it's indexable, and if I wanted to, I could switch that on for every user for a SSRd first load. I'm going to do something like that in one of my bigger projects.

**Comment by STRML** [comment]
Thanks, I think that's a helpful idea.
